lib/Dialect/Moore/MooreOps.cpp

```diff
@@ -329,6 +329,49 @@ void ConstantOp::build(OpBuilder &builder, OperationState &result, IntType type,
 // NamedConstantOp
 //===----------------------------------------------------------------------===//
 
+static ParseResult parseParamValue(OpAsmParser &p, Attribute &value,
+                                   Type &resultType) {
+  APInt valueInt;
+  IntType typeInt;
+  auto valueLoc = p.getCurrentLocation();
+
+  if (p.parseInteger(valueInt))
+    return failure();
+
+  if (p.parseColon() || p.parseCustomTypeWithFallback(typeInt))
+    return failure();
+
+  // Extend or truncate the constant value to match the size of the type.
+  if (typeInt.getWidth() > valueInt.getBitWidth()) {
+    // sext is always safe here, even for unsigned values, because the
+    // parseOptionalInteger method will return something with a zero in the
+    // top bits if it is a positive number.
+    valueInt = valueInt.sext(typeInt.getWidth());
+  } else if (typeInt.getWidth() < valueInt.getBitWidth()) {
+    // The parser can return an unnecessarily wide result with leading
+    // zeros. This isn't a problem, but truncating off bits is bad.
+    unsigned neededBits = valueInt.isNegative() ? valueInt.getSignificantBits()
+                                                : valueInt.getActiveBits();
+    if (typeInt.getWidth() < neededBits)
+      return p.emitError(valueLoc, "constant out of range for result type ")
+             << typeInt;
+
+    valueInt = valueInt.trunc(typeInt.getWidth());
+  }
+
+  auto attrType = IntegerType::get(p.getContext(), typeInt.getWidth());
+  value = IntegerAttr::get(attrType, valueInt);
+  resultType = typeInt;
+  return success();
+}
```

Comment:
I'm wondering if we really want to duplicate this code from `moore.constant`. I have a feeling that things will be more maintainable if we just make `named_constant` accept an SSA value, and if we really need that to be a constant, to add a verifier that checks whether that SSA value's defining op is a `moore.constant` op. That will prevent quite a bit of duplication, and only `moore.constant` would have to know how to map a constant to the HW dialect.

Especially since we currently don't _need_ the `named_constant` to be a constant value per se: when lowering to HW, you're mapping it to an `hw.wire`, which is great! That can deal even with non-constant values for now, so it might make sense to just not bother checking whether the value is constant at all.

Reply:
There is an example:
```
module Foo;
  parameter p1 = 32;
  parameter p2 = p1;
endmdule
```
lower to moore IR
```
moore.module (){
  %0 = moore.constant 32 : i32
  %p1 = moore.named_constant %0 : i32
  %p2 = moore.named_constant %p1 : i32
}
```
then lower to core IR
```
hw.module (){
  %c32_i32 = hw.constant 32 : i32
  %p1 = hw.wire %c32_i32 : i32
  %p2 = hw.wire %p1 : i32 
}
```
I think the above is what @fabianschuiki means.
We needn't lower `moore.named_constant` into `hw.para.value`, and also don't need `parseParamValue` that looks like parse `moore.constant`. And this lowering route can only lower `moore.constant` into `hw.constant` one-to-one.

Reply:
Thanks @fabianschuiki  @hailongSun2000 , I got it and updated my code. plz have a look =)